micronota/db/uniref.py

```diff
@@ -0,0 +1,210 @@
+# Copyright (c) 2015--, micronota development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from os.path import join, expanduser
+from sqlite3 import connect
+import shutil
+
+from skbio import read, Sequence
+
+from ..util import _overwrite
+
+
+def prepare_db(out_d, prefix='uniref100', force=False,
+               server='ftp.uniprot.org',
+               path='pub/databases/uniprot/uniref/uniref100',
+               fasta='uniref100.fasta.gz'):
+    '''Prepare reference database for UniRef.
+
+    Notes
+    -----
+    This function creates the following files:
+
+    * ``uniref100_sprot_archaea.dmnd``
+    * ``uniref100_sprot_bacteria.dmnd``
+    * ``uniref100_sprot_viruses.dmnd``
+    * ``uniref100_sprot_other.dmnd``
+    * ``uniref100_trembl_archaea.dmnd``
+    * ``uniref100_trembl_bacteria.dmnd``
+    * ``uniref100_trembl_viruses.dmnd``
+    * ``uniref100_trembl_other.dmnd``
+    * ``uniref100_other.dmnd``
+
+    * ``uniprotkb.db``
+    * ``uniref_100_90_50.db``: the map between UniRef100, UniRef90, and UniRef50.
+    '''
+
+    prepare_metadata(join(expanduser('~'), 'uniref', 'uniprot_sprot.dat.gz'),
+                     '%s.db' % join(out_d, prefix))
+    prepare_metadata(join(expanduser('~'), 'uniref', 'uniprot_trembl.dat.gz'),
+                     '%s.db' % join(out_d, prefix))
+    sort_uniref('%s.db' % join(out_d, prefix), join(expanduser('~'), 'uniref', 'uniref100.fasta.gz'), out_d)
+
+def create_map(fp):
+    '''Map between UniRef100, UniRef90, and UniRef50.
+    '''
+
+
+def sort_uniref(db_fp, uniref_fp, out_d):
+    '''Sort UniRef sequences into different partitions.'''
+
+    fn = ['%s_%s' % (i, j) for i in ['sprot', 'trembl']
+          for j in ['Bacteria', 'Archaea', 'Viruses', 'other']]
+    files = {f: open(join(out_d, 'uniref100_%s.fasta' % f), 'w') for f in fn}
+    files['_other'] = open('uniref100__other.fasta', 'w')
+    with connect(db_fp) as conn:
+        cursor = conn.cursor()
+        for seq in read(uniref_fp, format='fasta', constructor=Sequence):
+            id = seq.metadata['id']
+            id = id.replace('UniRef100_', '')
+            group = ['', 'other']
+            cursor.execute(
+                '''SELECT key, val FROM metadata
+                   WHERE id = ? AND key IN (?,?)''',
+                (id, 'KD', 'RE'))
+            for row in cursor.fetchall():
+                key, val = row
+                if key == 'KD':
+                    if val in ['Bacterial', 'Archaea', 'Viruses']:
+                        group[1] = val
+                if key == 'RE':
+                    group[0] = val
+            seq.write(files['_'.join(group)])
+    conn.commit()
+    for f in files:
+        files[f].close()
+
+
+def prepare_metadata(in_fp, fp, overwrite=False):
+    '''
```

Comment:
Parameters section?

Also, I'm beginning to see a pattern here.  These database preparation classes all have `prepare_db` and `prepare_metadata`.  

What do you think?  Do you think it would be worthwhile to enforce this abstraction?  I don't have a strong preference either way.